docs/benchmarking.md

```diff
@@ -0,0 +1,54 @@
+# Benchmarking
+
+## Why benchmark
+
+To encourage adoption we need Thrift.Net to be fast and compact. This will be
+particularly important for the client runtime but is also relevant for the compiler.
+
+## What to benchmark
+
+As a minimum we would like to benchmark on the following metrics:
+
+* execution time - how long the operation takes.
+* memory usage - how much memory is used by the operation. In a garbage
+collected environment this may also affect execution time.
+* memory retention - how much memory is retained after the operation is
+completed. This can be indicative of memory leaks.
+
+Since we are only currently working on the compiler the first priority will be
+to build a benchmarking
+suite suitable for the compiler. This will then help inform the benchmarking for
+the client runtime.
+
+### Against Apache
+
+We want to compare Thrift.Net against the Apache Thrift implementation. Since
+the Apache compiler is an executable we should benchmark against a generated
+executable of Thrift.Net to allow a fair comparison.
+
+One drawback of this approach is that it will probably only allow us to measure
+and compare execution time.
+
+### Against itself
+
+As Thrift.Net is being developed we also need to continue to evaluate it's
+performance against itself. This will ensure we don't have any performance
+regressions. With this approach we can also use a more specialized benchmarking
+tool that will allow measurements of memory usage. For this we will use [BenchmarkDotNet](https://github.com/dotnet/BenchmarkDotNet).
+
+## When we should benchmark
+
+To help prevent performance regressions we should aim to run the benchmarks on
+every commit as part of the code review process. It may be possible to add a
+comment to each pull request with details of the benchmarking suite - if so we
```

Comment:
I wonder if we can also compare it against the last successful build results so that if performance degrades by more than X% we can fail a check.